### ISSUE TITLE
in the ciou equa., the deviation of v wrt w and h for the scale factor sqrt(w**2+h**2) seems wrong

### ISSUE THREAD

**Issue by fjzhangcr** [Zzh-tju/DIoU#19]
hi, reading your aweson paper recently.
i mannually calc. the deviation of v wrt w and h.
maybe i am wrong about the deviation calculation, but it seems that the dv/dw should have a negative sign, and the dv/dh should not.
your deviation below:
![image](https://user-images.githubusercontent.com/102020341/168467219-06e46476-8b34-4499-a5f3-2f04ec87ba35.png)

my mannual calculation for deviation of v wrt w below:
![image](https://user-images.githubusercontent.com/102020341/168467270-4de5b88c-a2d5-44ed-9be7-4e8761bf14ea.png)

what did i miss?~~~thanks!

**Comment by fjzhangcr** [comment]
confusing~~~my alteration of ciou works well in the first epoch, but model converges worse in the following epochs...

**Comment by Zzh-tju** [comment]
Multiply the numerator and the denominator by h

**Comment by fjzhangcr** [comment]
yes, of course, i mean the negative sign~~~~the deviation of v w.r.t. w,  the manual caculation has a negative sign, your don't, would you please tell me why?

**Comment by Zzh-tju** [comment]
the negative sign depends on the difference between the two aspect ratios.

if w_gt/h_gt - w_pred/h_pred > 0，then dv/dw > 0 so as to enlarge the width of pred box, and vice versa. The custom gradient follow such principle. BTW, the network is updated by the negative gradient in pytorch.

Nevertheless, the custom gradient of v is now removed in the pytorch implementation.

**Comment by fjzhangcr** [comment]
something went wrong~~~
Quote:if w_gt/h_gt - w_pred/h_pred > 0，then dv/dw > 0 so as to enlarge the width of pred box, and vice versa. 
MyOpinion: i don't think so...when function y is a Concave function, when  dy/dx > 0, we must decrease x to get a lower y. it is a general rule. when applying to the v euqation, when dv/dw>0 we must decrease w, not enlarge it~~~i am pretty sure
BTW, would you please check the euqation 12 in your paper, i am pretty sure the deviation of v w.r.t. w lack of a negative sign~~~

**Comment by Zzh-tju** [comment]
There is no need for the custom gradient to be consistent with rigorous mathematical derivative. As you say, 
> when dy/dx > 0, we must decrease x to get a lower y.

the problem is, how to define dy/dx.

If pred=3, target=6, loss=(target-pred)^2, then by mathematical derivative, d(loss)/d(pred)=2*(target-pred)*(-1)=2(pred-target)<0. It is obvious that it runs counter to the idea that we want to increase pred. At this time, you must customize the gradient d(loss)/d(pred) to make pred increase.

Back to the loss of aspect ratio v, if the target aspect ratio > pred aspect ratio, it tells us that we should increase the width of the predicted bbox and decrease the height so as to reduce the difference of the two aspect ratios. 

Now, in the autograd deep learning framework, like pytorch, we can directly use the CIoU loss without the custom gradients.

**Comment by fjzhangcr** [comment]
`If pred=3, target=6, loss=(target-pred)^2, then by mathematical derivative, d(loss)/d(pred)=2*(target-pred)*(-1)=2(pred-target)<0. It is obvious that it runs counter to the idea that we want to increase pred. At this time, you must customize the gradient d(loss)/d(pred) to make pred increase.`
in equation 12, assume w_pred=3,w_gt=6, it is obvious that w_pred should increase. according the gradient decent algorithm, the d_loss/dw should be a negative number, so as to increas w for a lower loss~~~ let us check the equation12.
8/(pi**2) is positive, arctan(w_gt/h_gt)>arctan(w_pred/h_pred), h/(w**2+h**2) is positive. All 3 factor is all positive, how can we obtain a negative dv/dw~~~?
Unless, the equation 12 refers to a different term or definition or some customer anything, ~~~but it is absolutly not the traditional definition of derivation.

**Comment by Zzh-tju** [comment]
Eq.12 describes the gradient directions we want. Implementation is based on that principle.

**Comment by fjzhangcr** [comment]
> Eq.12 describes the gradient directions we want. Implementation is based on that principle.

yes, that make sense~~BTW, what will you do with the 

> **gradient directions we want** ( use GDWW for short )


- Option A: the **gradient directions we want ( use GDWW for short )**  is used to add to the w_pred which means : **new_w_pred = w_pred + lr*GDWW)** or 
- Option B: the **gradient directions we want ( use GDWW for short )**  is used to subtract from the w_pred which mens: **new_w_pred = w_pred - lr*GDWW)?** 
in which lr for learning rate

take the w_gt=6, w_pred=3 as an example, the w_pred should be increase obviously~~ in this scenario, equation 12 about w is what you say  **gradient directions we want ( use GDWW for short )**  which is obviously positive.
My guess: your choice is Option A, because the new_w_pred will increase which fits the common sense(w_pred=3 should increase to reach the w_gt=6)?
Am I right? It is important because in tensorflow the gradient result is used to assign-sub from the variable~~~two options should choose carefully according to gradicent decent implementation machinism.

**Comment by Zzh-tju** [comment]
The gradient directions depend on the difference between the two aspect ratios.

If w_pred=3，h_pred=3，w_gt=6，h_gt=12，then ar_pred=1，ar_gt=0.5，we want w_pred to decrease and h_pred to increase so as to fit the aspect ratios.

**Comment by fjzhangcr** [comment]
yes, let me make it clear: assume h_true equals h_pred, and two bboxes‘s centers coincident, when w_pred=3,w_gt=6, then we have loss=1-iou+rho_term+alpha*v=1-0.5+0+alpha*v.... what i posted was based on this assumption~~
let me get back to my two options:
the key question is
 **what will you do with the gradient directions we want ( use GDWW for short ).**

- Option A: add the gradient directions we want ( use GDWW for short ) to the w_pred:  : new_w_pred = w_pred + lr*GDWW) 
- Option B subtract the gradient directions we want ( use GDWW for short ) from the w_pred: new_w_pred = w_pred - lr*GDWW)?
- in which lr for learning rate

**Comment by Zzh-tju** [comment]
it depends on the deep learning framework you use. you can see the code YOLOv3 (https://github.com/Zzh-tju/DIoU-darknet/blob/master/src/box.c#L523).

And if you use TensorFlow, you don't need to use the custom gradient, just calculate the aspect ratio loss v, and multiply by alpha (no gradient backward).

**Comment by fjzhangcr** [comment]
[	float ar_dw=8/(PI*PI)*(atan(ar_gt)-atan(ar_pred))*pred.h;](https://github.com/Zzh-tju/DIoU-darknet/blob/95a97f760d7c358e592645239032bdc2fb55b35c/src/box.c#L523)
...略。。
[	p_dw += (2*Cw*dCw_dw+2*Ch*dCh_dw)*S / (C * C) + alpha * ar_dw;](https://github.com/Zzh-tju/DIoU-darknet/blob/95a97f760d7c358e592645239032bdc2fb55b35c/src/box.c#L529)

you use option A~~ which is assign_add to the w_pred while tensorflow auto-grad mechinism use assign_sub~~ I know what the mis-understanding is~~~ some framework(such as TF)  subtract the gradient frome the prediction, some others(like your yolov3 C version)  add the gradient to the predection~~
